### ISSUE TITLE
Question about slicing, --multi-property

### ISSUE THREAD

**Issue by zhuyutian57** [esbmc/esbmc#1960]
I am curious about the idea about slicing for `--multi-property`. Though the purpose of `--multi-property` is to check each `ASSERT` in `eq` system, it seems that the slicing is not enough.

In `job_function`, each `ASSERT` will be checked independently. See the following codes.

```c++
    // Set up the current claim and disable slice info output
    claim_slicer claim(i, false, is_goto_cov, ns);
    claim.run(local_eq.SSA_steps);

    // Drop claims that verified to be failed
    // we use the "comment + location" to distinguish each claim
    // to avoid double verifying the claims that are already verified
   ......

    // Slice
    symex_slicet slicer(options);
    slicer.run(local_eq.SSA_steps);
```

Firstly, `claim_slicer` will set the $i^{th}$ `ASSERT` to be checked. Specifically, `claim_slicer` will set the $i^{th}$ `ASSERT`to be checked(`step.ignored = false`) while the others is set to be ignored(`step.ignored = true`).

Secondly, the `symex_slicet` will do the slicing for the whole constructed `eq` system.

However, in the second step, `symex_slicer` will do the slicing without ignoring the other `ASSERT`s. But in the next step, encoding to SMT formulas, only the $i^{th}$ `ASSERT` is used for encoding. Doing this will introduce redundant some statements.

Is there a bug? Or I misunderstand some designs for `Array` encoding?
For example, checking memory leak for the below program.

```c++
    int * data = malloc(2 * sizeof(int));
    int * data2 = malloc(2 * sizeof(int));
    free(data2);
    int i = *(data + 2);
``` 

The translation from `GOTO-C` to `eq` system will insert two memory leak assertions in the end. One for `data` and the another is for `data2`. Ideally, `data2` should not be included when we checking memory leak for `data`. Am I right?

However, `symex_slicet` will take all `ASSERT` into account. See below codes in `symex_slicet`. The `run_on_step` will handle all `ASSERT`. That will take all symbols into account.

```c++
  bool run(symex_target_equationt::SSA_stepst &eq) override
  {
    sliced = 0;
    fine_timet algorithm_start = current_time();
    int cnt = 0;
    for (auto &step : boost::adaptors::reverse(eq))
      run_on_step(step);
    fine_timet algorithm_stop = current_time();
    log_status(
      "Slicing time: {}s (removed {} assignments)",
      time2string(algorithm_stop - algorithm_start),
      sliced);
    return true;
  }
```

Thus, if we ignore the `ASSERT`s that have been ignored, we get the desired VCC.

```c++
    .......
    for (auto &step : boost::adaptors::reverse(eq)) {
      if (step.is_assert() && step.ignore) continue;
      run_on_step(step);
    }
    ......
```

Is there is a bug? Or I misunderstand some designs for `Array` encoding?

**Comment by rafaelsamenezes** [comment]
You are correct. The ignored assertions shouldn't affect the formula. However, I would claim that every ignored instruction should be skipped.

```
for (auto &step : boost::adaptors::reverse(eq)) {
      if (step.ignore) continue;
      run_on_step(step);
    }
```

**Comment by zhuyutian57** [comment]
Thanks. You are right. This will simplify the encoding for checking one assertion.

**Comment by lucasccordeiro** [comment]
Closed via #1961.
